sdk/include/opentelemetry/sdk/metrics/state/sync_metric_storage.h

```diff
@@ -101,8 +108,112 @@ class SyncMetricStorage : public MetricStorage, public WritableMetricStorage
                opentelemetry::common::SystemTimestamp collection_ts,
                nostd::function_ref<bool(MetricData &)> callback) noexcept override
   {
-    MetricData data;
-    if (callback(data))
+    opentelemetry::common::SystemTimestamp last_collection_ts = sdk_start_ts;
+    auto aggregation_temporarily = collector->GetAggregationTemporarily();
+
+    // Add the current delta metrics to `unreported metrics stash` for all the collectors,
+    // this will also empty the delta metrics hashmap, and make it available for
+    // recordings
+    std::shared_ptr<AttributesHashMap> delta_metrics = std::move(attributes_hashmap_);
+    attributes_hashmap_.reset(new AttributesHashMap);
+    for (auto &col : collectors)
+    {
+      unreported_metrics_[col.get()].push_back(delta_metrics);
+    }
+
+    // Get the unreported metrics for the `collector` from `unreported metrics stash`
+    // since last collection, this will also cleanup the unreported metrics for `collector`
+    // from the stash.
+    auto present = unreported_metrics_.find(collector);
+    if (present == unreported_metrics_.end())
+    {
+      // no unreported metrics for the collector, return.
+      return true;
+    }
+    auto unreported_list = std::move(unreported_metrics_[collector]);
```

Comment:
Use the variable present for move to avoid another lookup on `unreported_metrics_`?

Reply:
Good point. Have changed accordingly.